### DISCUSSION TITLE
DUCKDB_API C++ build on Windows

### DISCUSSION THREAD

**Discussion by ghost** [duckdb/duckdb#2774] [Q&A]
Hi,

This is an excellent project!

I noticed this code in `duckdb.h`

```#ifdef _WIN32  
#ifdef DUCKDB_BUILD_LIBRARY  
#define DUCKDB_API __declspec(dllexport)  
#else  
#define DUCKDB_API __declspec(dllimport)  
#endif  
#else  
#define DUCKDB_API  
#endif
```

Why is only the shared DLL library option available here for ``_WIN32``? I was wondering if there could be a static build option here? It seems it's an easy fix, just have a flag to allow ``#define DUCK_API`` (without any ``__declspec`` etc) for any platform including ``_WIN32``? I think this would make including the amalgamation build directly into a project more straightforward. 

Thanks!

**Comment by Mytherin** [comment]
Sure, that sounds good to me, happy to review a PR!

**Reply by ghost** [reply]
Hi Mytherin, 

If you support compiling and using DuckDB as a static library, then I would suggest introducing a new pre-processor define for the DuckDB project to indicate whether or not it is a static build. Arrow has ``ARROW_STATIC`` so maybe it could be called ``DUCKDB_STATIC``?

Unfortunately, I don't understand enough about DuckDB to submit a PR, at the moment. I am confused because DUCKDB_API is defined in two places, in ``duckdb.h`` as mentioned above as well as in ``common/winapi.hpp``, and I'm not sure if there is some reason for the duplication:

```
#ifndef DUCKDB_API
#ifdef _WIN32
#if defined(DUCKDB_BUILD_LIBRARY) && !defined(DUCKDB_BUILD_LOADABLE_EXTENSION)
#define DUCKDB_API __declspec(dllexport)
#else
#define DUCKDB_API __declspec(dllimport)
#endif
#else
#define DUCKDB_API
#endif
#endif

#ifdef DUCKDB_BUILD_LOADABLE_EXTENSION
#ifdef _WIN32
#define DUCKDB_EXTENSION_API __declspec(dllexport)
#else
#define DUCKDB_EXTENSION_API
#endif
#endif
```

**Reply by Mytherin** [reply]
The duplication is because duckdb.h is shipped as a standalone header to clients that can use it together with the duckdb library. When shipped they don’t have access to the rest of the headers. Just duplicate the changes you make in both locations.

**Comment by virtualsafety** [comment]
when i comple the c++ example code https://github.com/duckdb/duckdb/blob/master/examples/embedded-c%2B%2B-windows/cppintegration.cpp, i got  lots of similar errors like the follwing:
duckdb.hpp:22076:26: error: function 'duckdb::idx_t duckdb::BaseAppender::CurrentColumn()' definition is marked dllimport

**Comment by taniabogatsch** [comment]
It has been some time, but a PR is here to add the respective option: https://github.com/duckdb/duckdb/pull/13744.
